### PR TITLE
Update .NET samples to .NET 8 and Minimal API

### LINKED ISSUE #387: Convert .NET API to Minimal APIs

### PR DESCRIPTION
**Pull request by isaacrlevin**
Fix #387 
- Update all .NET samples and tests from .NET 6 to .NET 8 (latest LTS)
- Update all .NET samples to use Minimal APIs

### CODE REVIEW

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (davidfowl) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:7]: The formatting and spacing in this file is off.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (davidfowl) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:5]: @captainsafia can you review this please?

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (captainsafia) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:16]: `TypedResults.Ok` implies a 200 status code so you shouldn't need to add this.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (captainsafia) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:9]: Recommend targeting `IEndpointRouteBuilder` with this and returning the builder object to support chaining other calls at the invocation site.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (captainsafia) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:12]: No need for the explicit `[FromQuery]` attribute here. It's implied for parsable parameter types that don't have an associated route parameter.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (captainsafia) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:20]: Similar to `[FromQuery]`, no need for `[FromBody]` here since it is implied for complex types that aren't registered in DI.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (captainsafia) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:29]: Recommend using `TypedResults` here. Similarly, you won't need `Produces(201)` if you use the `TypedResults` variant.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (captainsafia) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:31]: Since `WithOpenApi` is called repeatedly, you may consider moving it to the `Program.cs` after the `MapTodoEndpoints` call. You'll need to update the signature of this method as mentioned above.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (captainsafia) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:107]: Perhaps we should implement a copy constructor on `CreateUpdateTodoItem` and make it a class?

**Review by captainsafia — COMMENTED** [review]
Posted some initial comments inline. In general:

- Use `TypedResults` over `Results` where applicable
- Move repeated `WithOpenApi` calls to a single call on `MapTodoEndpoints`
- Avoid explicit `[FromX]` attributes where the implicit behavior sufficies
- Consider using a route group to capture the `/lists` prefix

**Comment by isaacrlevin** [comment]
@captainsafia let me know when you are done with your review and I will do a pass. Appreciate the help!

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:107]: Not sure how that would work as you eventually need to convert back to TodoItem for the datasource. Unless I am misunderstanding what you are suggesting

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (davidfowl) [inline templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:76]: Can we change the _ convention, it's not C# like to use _ in method paramters. Also the naming `list_id` is not idiomatic either. These should be:

```suggestion
        public static async Task<IResult> GetListItems(ListsRepository repository, string listId, int? skip = null, int? batchSize = null)

```

File: templates/todo/api/csharp-sql/TodoEndPointsExt.cs
Comment (davidfowl) [inline templates/todo/api/csharp-sql/TodoEndPointsExt.cs:23]: nit: Spacing
```suggestion

        public static async Task<Ok<IEnumerable<TodoList>>> GetLists(ListsRepository _repository, int? skip = null, int? batchSize = null)

```

File: templates/todo/api/csharp-sql/TodoEndPointsExt.cs
Comment (davidfowl) [inline templates/todo/api/csharp-sql/TodoEndPointsExt.cs:158]: nit:
```suggestion
```

File: templates/todo/api/csharp/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp/Program.cs:11]: azd builds the infra for you than deploys the app all in one step. In order for that to work, azd uses bicep and outputs to get things like keys and connection strings. In this case, the Cosmos connection string is put in KeyVault with a particular key. So no settings are actually set locally at all. All the azd templates work in this way

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs:76]: Sure, I just followed the convention the sample already used

File: templates/todo/api/csharp/Program.cs
Comment (davidfowl) [reply templates/todo/api/csharp/Program.cs:11]: I see this is just consuming environment variables. Got it. Future versions will make this more idiomatic. Still, you don't need 2 layers of lookup do you?

**Review by skydiver1980 — APPROVED** [review]

File: templates/todo/api/csharp/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp/Program.cs:11]: The original template did it this way and I just left as is. I assume this is just an extra privacy measure.
- db connection string out in key vault with a key value
- the value of the key is added to app environment variables
- app gets value of db connection string in key vault


I can do whatever you think is the best approach

**Comment by skydiver1980** [comment]
The ball is in your court, make the decision you think best lol, I have
faith.

File: templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp-cosmos-sql/TodoEndPointsExt.cs]: resolving per our chat on Teams

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/internal/appdetect/testdata/dotnet/dotnettestapp.csproj
Comment (vhvb1989) [inline cli/azd/internal/appdetect/testdata/dotnet/dotnettestapp.csproj:5]: We should update project's Readme, like https://github.com/Azure/azure-dev/blob/main/templates/todo/projects/csharp-cosmos-sql/README.md?plain=1#L43

Comment (isaacrlevin) [reply cli/azd/internal/appdetect/testdata/dotnet/dotnettestapp.csproj:5]: done

File: templates/todo/api/csharp/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp/Program.cs:10]: Looking into this more, the change needed to accomodate this ask would involve me having to update the core bicep files for the templates and have to update all the templates in kind. @savannahostrowski @jongio would appreciate your thoughts here.

**Review by ellismg — APPROVED** [review]

File: templates/todo/api/csharp/Program.cs
Comment (jongio) [reply templates/todo/api/csharp/Program.cs:10]: Please open an issue for this.  As a pattern we store the connection string in keyVault and then lookup that value.  The setting here is the key name and we aren't ready to make that change as part of this PR.  We'll need to discuss.

File: templates/todo/api/csharp-cosmos-sql/Program.cs
Comment (jongio) [inline templates/todo/api/csharp-cosmos-sql/Program.cs:10]: Why not using DefaultAzureCredential and endpoint?

Comment (jongio) [reply templates/todo/api/csharp-cosmos-sql/Program.cs:10]: In all cases we want to favor RBAC over connection string usage.

File: templates/todo/api/csharp/Program.cs
Comment (jongio) [inline templates/todo/api/csharp/Program.cs:10]: You should be using Endpoint and DAC, not connection string

**Review by jongio — CHANGES_REQUESTED** [review]

File: templates/todo/api/csharp-cosmos-sql/Program.cs
Comment (jongio) [inline templates/todo/api/csharp-cosmos-sql/Program.cs:10]: I think you need the CosmosClientOPtions as well.

File: templates/todo/api/csharp-sql-func/ListsFunctions.cs
Comment (jongio) [inline templates/todo/api/csharp-sql-func/ListsFunctions.cs:10]: Can you verify that all c# files in our repo across all templates follow this convention?

File: templates/todo/api/csharp/Program.cs
Comment (jongio) [inline templates/todo/api/csharp/Program.cs:10]: Reason for switching from Mongo to Cosmos client?  @karolz-ms might know why we went with MongoClient  initially.

**Review by jongio — CHANGES_REQUESTED** [review]
Thanks for doing this.  Let's look at each place where you are using connection string and see if we can use RBAC instead.

File: templates/todo/api/csharp/Program.cs
Comment (vhvb1989) [reply templates/todo/api/csharp/Program.cs:10]: The definition and output from the /core modules are agnostic to the language/app that is hosting (net/java/js/py).  I would not suggest updating the /core modules.

But, within each project, we have an `api.bicep` and `web.bicep` next to the `main.bicep`.  These modules are actually the ones which are aware of the language and specific requirements for the service to be hosted.  We can make this modules to output the .net idiomatic names for env-vars and then the App can use it.

Comment (isaacrlevin) [reply templates/todo/api/csharp/Program.cs:10]: I think the ask here is to update the KEY of the connection string stored in KV to use the .NET naming convention. This is something that would need to be done in the CORE folders (the KEY is actually defined [here](https://github.com/Azure/azure-dev/blob/70655e6638a6225020762cc8c9c3315b940cc702/templates/common/infra/bicep/core/database/cosmos/mongo/cosmos-mongo-db.bicep#L8) in this case). Unless I am mistaken @davidfowl

File: templates/todo/api/csharp-cosmos-sql/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp-cosmos-sql/Program.cs:10]: @jongio This already existed in the sample. More than happy to update but didn't want to update too much.

File: templates/todo/api/csharp/Program.cs
Comment (vhvb1989) [reply templates/todo/api/csharp/Program.cs:10]: Oh, I see.
So, the change is to set the key as an input param for the bicep module, keeping the actual value as the default value.
Then, from main.bicep, you can set the name to use as key (or secret) to keep the conn-string

Comment (jongio) [reply templates/todo/api/csharp/Program.cs:10]: You can override the key name in you main.bicep. 

Here's the setting to override in .NET templates.

https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/database/cosmos/mongo/cosmos-mongo-db.bicep#L8

File: templates/todo/api/csharp-cosmos-sql/Program.cs
Comment (jongio) [reply templates/todo/api/csharp-cosmos-sql/Program.cs:10]: If you don't want to take it in this PR, then let's keep as-is, use RBAC if already using RBAC, use KEY if using key, then please create an issue to get them all updated to use RBAC.

File: templates/todo/api/csharp/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp/Program.cs:10]: Got it, so I can pass it whatever I want in the "local" bicep template that is aligned to the sample? Let me give that a try to see if it works.

**Review by danieljurek — COMMENTED** [review]
Changes to `eng/pipelines/release-cli.yml` look good to me at this time. I want to look a little more closely at `cli/Dockerfile` before signing off.

File: templates/todo/api/csharp-cosmos-sql/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp-cosmos-sql/Program.cs:10]: Doing in another PR per @jongio

File: templates/todo/api/csharp/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp/Program.cs]: Removing this template in another PR per @jongio

File: cli/Dockerfile
Comment (danieljurek) [inline cli/Dockerfile:21]: Why are `gcc`, `g++`, and `make` needed?

Comment (isaacrlevin) [reply cli/Dockerfile:21]: No idea, @vhvb1989 can you provide some context?

File: cli/Dockerfile
Comment (danieljurek) [inline cli/Dockerfile:23]: Version differences: 

| Tool | Latest container version | Version in PR | 
| ---- | ------------------------ | ------------- | 
| Debian | 11 | 12 | 
| dotnet | 6.0.417 | 8.0.100 | 
| node | 18.17.1 | 18.13.0 | 
| npm | 9.6.7 | 9.2.0 | 
| python | 3.9.2 | 3.11.2 | 
| gcc/g++ | n/a | 12.2.0 | 
| make | n/a | 4.3 | 

* Linux distro upgrade -- This should not be a major concern as it is the base image for the .NET container. Looks like 8.0 is not shipping on MCR for `bullseye` (aka Debian 11). OK ✅
* `dotnet` -- This is the latest released version and comes with the container. OK ✅
* `node`/`npm` -- These differ only by minor version which won't impact `azd` per se but in the case of node/npm the versions are going backward. Do we care? If so, use instructions here for onboarding later versions: https://github.com/nodesource/distributions#installation-instructions
* `python` -- Minor version upgrade to Python. Not expected to cause problems as we do our template dev container testing on Python 3.10 which is one minor version ahead.

File: cli/Dockerfile
Comment (vhvb1989) [reply cli/Dockerfile:21]: This is after removing: 

`&& curl -fsSL https://deb.nodesource.com/setup_18.x | bash - \`

We were using https://deb.nodesource.com/setup_18.x to install node, but it is not deprecated.

I saw a message at the end of the script about installing c/c++ for building native:

image

Comment (vhvb1989) [reply cli/Dockerfile:21]: I'll update by using the installation instructions from https://github.com/nodesource/distributions

Comment (vhvb1989) [reply cli/Dockerfile]: updated as it should now.
Thank you @danieljurek

File: templates/todo/api/csharp/Program.cs
Comment (isaacrlevin) [reply templates/todo/api/csharp/Program.cs]: After chatting with @davidfowl, we are going to hold off on this ask. Resolving convo

**Comment by isaacrlevin** [comment]
@vhvb1989 @jongio @danieljurek all the work is done in this PR. Let me know if there is anything additional needed.

**Comment by vhvb1989** [comment]
@jongio , `test- templates` pipeline is passing. This PR is ready, can you unblock?

**Comment by isaacrlevin** [comment]
@jongio I reverted back to RBAC in one sample. Should be good now

**Comment by jongio** [comment]
> I reverted back to RBAC in one sample. Should be good now

Thanks.  I've asked @v-xuto to run regression tests on this since there are so many changes.

**Comment by jongio** [comment]
@v-xuto Here they are, sorry about that, had issue with the pipeline https://github.com/Azure/azure-dev/pull/3017#issuecomment-1841308811

**Review by jongio — APPROVED** [review]
Please wait until testing is complete before merging

**Comment by isaacrlevin** [comment]
Fix #3086

**Comment by isaacrlevin** [comment]
Fix #3090

**Comment by isaacrlevin** [comment]
Synced with upstream

**Comment by vhvb1989** [comment]
Fix: https://github.com/Azure/azure-dev/issues/3089

**Pull request merged**
